--- a/download.lst	2024-09-07 13:27:30.513193986 +0200
+++ b/download.lst	2024-09-07 13:28:39.455198995 +0200
@@ -650,8 +650,8 @@
 # three static lines
 # so that git cherry-pick
 # will not run into conflicts
-ZXCVBN_C_SHA256SUM := 77d6c6ecb35952a8d8ce7f736b7a2bf466275c48210e309b73782d6b7e84dffd
-ZXCVBN_C_TARBALL := zxcvbn-c-2.5.tar.gz
+ZXCVBN_C_SHA256SUM := 33d3ba80edab17cd730b4417370ba8e01923e60a1d5325f8ce458cdf951a2a37
+ZXCVBN_C_TARBALL := zxcvbn-c-b8a3d0b5992d1f50aa7559886f7e2ffc6740e2b7.tar.gz
 # three static lines
 # so that git cherry-pick
 # will not run into conflicts
