--- a/lib/grm/src/grm/backtrace.c
+++ b/lib/grm/src/grm/backtrace.c
@@ -3,7 +3,7 @@
 #endif
 
 /* FreeBSD excluded due to BinaryBuilder issues */
-#if (defined(__unix__) || defined(__APPLE__)) && !defined(__EMSCRIPTEN__) && !defined(__FreeBSD__)
+#if (defined(__GLIBC__) || defined(__APPLE__)) && !defined(__EMSCRIPTEN__) && !defined(__FreeBSD__)
 #define BACKTRACE_AVAILABLE 1
 #endif
 
--- a/lib/gks/io.c
+++ b/lib/gks/io.c
@@ -11,7 +11,7 @@
 #include <io.h>
 #endif
 
-#if defined(cray) || defined(__SVR4) || defined(_WIN32)
+#if defined(cray) || defined(__SVR4) || defined(_WIN32) || !defined(__GLIBC__)
 #include <fcntl.h>
 #else
 #include <sys/file.h>
