--- a/make/autoconf/libraries.m4
+++ b/make/autoconf/libraries.m4
@@ -43,9 +43,11 @@ AC_DEFUN_ONCE([LIB_DETERMINE_DEPENDENCIES],
   if test "x$OPENJDK_TARGET_OS" = xwindows || test "x$OPENJDK_TARGET_OS" = xmacosx; then
     # No X11 support on windows or macosx
     NEEDS_LIB_X11=false
+  elif test "x$ENABLE_HEADLESS_ONLY" = xtrue; then
+    # No X11 support needed when building headless only
+    NEEDS_LIB_X11=false
   else
-    # All other instances need X11, even if building headless only, libawt still
-    # needs X11 headers.
+    # All other instances need X11
     NEEDS_LIB_X11=true
   fi
 
--- a/make/modules/java.desktop/lib/AwtLibraries.gmk	2024-07-09 10:07:16.000000000 +0200
+++ b/make/modules/java.desktop/lib/AwtLibraries.gmk	2024-08-29 15:46:17.583713783 +0200
@@ -110,6 +110,10 @@
 
 LIBAWT_CFLAGS += -DMLIB_NO_LIBSUNMATH
 
+ifeq ($(ENABLE_HEADLESS_ONLY), true)
+    LIBAWT_CFLAGS += -DHEADLESS
+endif
+
 ifeq ($(call isTargetOs, windows), true)
   LIBAWT_CFLAGS += -EHsc -DUNICODE -D_UNICODE
   ifeq ($(call isTargetCpuBits, 64), true)
--- a/src/java.desktop/unix/native/common/awt/utility/rect.h	2024-07-09 10:07:16.000000000 +0200
+++ b/src/java.desktop/unix/native/common/awt/utility/rect.h	2024-08-29 15:46:17.919710301 +0200
@@ -28,7 +28,7 @@
 #ifndef _AWT_RECT_H
 #define _AWT_RECT_H
 
-#ifndef MACOSX
+#if !defined(MACOSX) && !defined(HEADLESS)
 #include <X11/Xlib.h>
 typedef XRectangle RECT_T;
 #else
--- a/src/java.desktop/unix/native/common/java2d/opengl/GLXGraphicsConfig.h	2024-07-09 10:07:16.000000000 +0200
+++ b/src/java.desktop/unix/native/common/java2d/opengl/GLXGraphicsConfig.h	2024-08-29 15:46:17.919710301 +0200
@@ -26,16 +26,17 @@
 #ifndef GLXGraphicsConfig_h_Included
 #define GLXGraphicsConfig_h_Included
 
-#include "jni.h"
-#include "J2D_GL/glx.h"
-#include "OGLSurfaceData.h"
-#include "OGLContext.h"
-
 #ifdef HEADLESS
 #define GLXGraphicsConfigInfo void
 #define GLXCtxInfo void
+#define CAPS_EMPTY 0L
 #else /* HEADLESS */
 
+#include "jni.h"
+#include "J2D_GL/glx.h"
+#include "OGLSurfaceData.h"
+#include "OGLContext.h"
+
 /**
  * The GLXGraphicsConfigInfo structure contains information specific to a
  * given GLXGraphicsConfig (visual).  Each AwtGraphicsConfigData struct
--- a/src/java.desktop/unix/native/common/java2d/opengl/GLXSurfaceData.c	2024-07-09 10:07:16.000000000 +0200
+++ b/src/java.desktop/unix/native/common/java2d/opengl/GLXSurfaceData.c	2024-08-29 15:46:17.919710301 +0200
@@ -27,9 +27,12 @@
 
 #include "sun_java2d_opengl_GLXSurfaceData.h"
 
+#ifndef HEADLESS
 #include "OGLRenderQueue.h"
 #include "GLXGraphicsConfig.h"
 #include "GLXSurfaceData.h"
+#endif
+
 #include "awt_Component.h"
 #include "awt_GraphicsEnv.h"
 
--- a/src/java.desktop/unix/native/common/java2d/opengl/GLXSurfaceData.h	2024-07-09 10:07:16.000000000 +0200
+++ b/src/java.desktop/unix/native/common/java2d/opengl/GLXSurfaceData.h	2024-08-29 15:46:17.919710301 +0200
@@ -26,14 +26,14 @@
 #ifndef GLXSurfaceData_h_Included
 #define GLXSurfaceData_h_Included
 
-#include "J2D_GL/glx.h"
-#include "awt_p.h"
-#include "OGLSurfaceData.h"
-
 #ifdef HEADLESS
 #define GLXSDOps void
 #else /* HEADLESS */
 
+#include "J2D_GL/glx.h"
+#include "awt_p.h"
+#include "OGLSurfaceData.h"
+
 /**
  * The GLXSDOps structure contains the GLX-specific information for a given
  * OGLSurfaceData.  It is referenced by the native OGLSDOps structure.
--- a/src/java.desktop/unix/native/common/java2d/x11/X11SurfaceData.h	2024-07-09 10:07:16.000000000 +0200
+++ b/src/java.desktop/unix/native/common/java2d/x11/X11SurfaceData.h	2024-08-29 15:46:17.919710301 +0200
@@ -30,9 +30,9 @@
 
 #ifdef HEADLESS
 #include "GLXGraphicsConfig.h"
-#endif
-
+#else
 #include <X11/extensions/Xrender.h>
+#endif
 
 /**
  * This include file contains support declarations for loops using the
