--- a/replaygain/ffmpeginput.cpp	2024-06-23 16:39:45.041262172 +0200
+++ b/replaygain/ffmpeginput.cpp	2024-06-23 16:40:45.081253909 +0200
@@ -26,7 +26,7 @@
 #ifdef __cplusplus
 }
 #endif
-#include "ebur128.h"
+#include <ebur128.h>
 #include "ffmpeginput.h"
 #include <QByteArray>
 #include <QFile>
--- a/replaygain/mpg123input.cpp	2024-06-23 16:39:53.077927984 +0200
+++ b/replaygain/mpg123input.cpp	2024-06-23 16:41:31.221244801 +0200
@@ -15,7 +15,7 @@
 #ifdef __cplusplus
 }
 #endif
-#include "ebur128.h"
+#include <ebur128.h>
 #include "mpg123input.h"
 #include <QByteArray>
 #include <QFile>
--- a/replaygain/trackscanner.h	2024-06-23 16:40:05.987926443 +0200
+++ b/replaygain/trackscanner.h	2024-06-23 16:41:44.977908323 +0200
@@ -24,7 +24,7 @@
 #ifndef _SCANNER_H_
 #define _SCANNER_H_
 
-#include "ebur128.h"
+#include <ebur128.h>
 #include "jobcontroller.h"
 
 class Input;
