--- a/XADRegex.m
+++ b/XADRegex.m
@@ -141,7 +141,7 @@
 {
 	matches[0].rm_so=matchrange.location;
 	matches[0].rm_eo=matchrange.location+matchrange.length;
-	if(regexec(&preg,[currdata bytes],preg.re_nsub+1,matches,REG_STARTEND)==0)
+	if(regexec(&preg,[currdata bytes],preg.re_nsub+1,matches,0)==0)
 	{
 		matchrange.length-=matches[0].rm_eo-matchrange.location;
 		matchrange.location=(long)matches[0].rm_eo;
