diff --git a/plugins/rdp/CMakeLists.txt b/plugins/rdp/CMakeLists.txt
index da4798b1d6f4db59e3302d0d30b6d8069b5314cd..f24ece2ef7c7e7f21785106f4fd39b01ab1048ea 100644
--- a/plugins/rdp/CMakeLists.txt
+++ b/plugins/rdp/CMakeLists.txt
@@ -55,6 +55,8 @@ set(REMMINA_PLUGIN_RDP_SRCS
     rdp_monitor.h
     rdp_channels.c
     rdp_channels.h
+    rdp_web_auth.c
+    rdp_web_auth.h
     )
 
 add_definitions(-DFREERDP_REQUIRED_MAJOR=${FREERDP_REQUIRED_MAJOR})
@@ -65,9 +67,54 @@ add_library(remmina-plugin-rdp MODULE ${REMMINA_PLUGIN_RDP_SRCS})
 set_target_properties(remmina-plugin-rdp PROPERTIES PREFIX "")
 set_target_properties(remmina-plugin-rdp PROPERTIES NO_SONAME 1)
 
-include_directories(${REMMINA_COMMON_INCLUDE_DIRS} ${FreeRDP-Client_INCLUDE_DIR} ${FreeRDP_INCLUDE_DIR} ${WinPR_INCLUDE_DIR})
+include_directories(
+    ${REMMINA_COMMON_INCLUDE_DIRS}
+    ${FreeRDP-Client_INCLUDE_DIR}
+    ${FreeRDP_INCLUDE_DIR}
+    ${WinPR_INCLUDE_DIR})
 target_link_libraries(remmina-plugin-rdp
-    ${REMMINA_COMMON_LIBRARIES} freerdp-client)
+    ${REMMINA_COMMON_LIBRARIES}
+    freerdp-client)
+
+option(WITH_RDP_AUTH_AAD "Enable AAD authentication support" OFF)
+if(WITH_RDP_AUTH_AAD)
+    if(WITH_FREERDP3)
+        file(READ ${FreeRDP_INCLUDE_DIR}/freerdp/config.h FREERDP_CONFIG_CONTENT)
+        if("${FREERDP_CONFIG_CONTENT}" MATCHES "#define WITH_AAD")
+            set(FREERDP_HAS_AUTH_AAD TRUE)
+        else()
+            message(WARNING "FreeRDP is built without AAD authentication support")
+        endif()
+    else()
+        message(WARNING "FreeRDP 2 does not support AAD authentication")
+    endif()
+
+    find_required_package(WEBKIT2GTK)
+    if(NOT WEBKIT2GTK_FOUND)
+        message(WARNING "WebKit-GTK not found")
+    endif()
+
+    find_required_package(LIBSOUP)
+    if(NOT WEBKIT2GTK_FOUND)
+        message(WARNING "libsoup not found")
+    endif()
+
+    if(FREERDP_HAS_AUTH_AAD AND WEBKIT2GTK_FOUND AND LIBSOUP_FOUND)
+        message(STATUS "Enabling RDP AAD authentication.")
+        # we cannot undefine WITH_AAD defined in FreeRDP so we use our own macro
+        add_definitions(-DWITH_RDP_AUTH_AAD)
+        include_directories(
+            ${WEBKIT2GTK_INCLUDE_DIRS}
+            ${LIBSOUP_INCLUDE_DIRS})
+        target_link_libraries(remmina-plugin-rdp
+            ${LIBSOUP_LIBRARIES}
+            ${WEBKIT2GTK_LIBRARIES})
+    else()
+        message(WARNING "Disabling RDP AAD authentication.")
+    endif()
+else()
+    message(STATUS "Skipping RDP AAD authentication.")
+endif()
 
 if(CUPS_FOUND)
     add_definitions(-DHAVE_CUPS)
diff --git a/plugins/rdp/rdp_plugin.c b/plugins/rdp/rdp_plugin.c
index 127198a164c66da07dfbcc82a75aaf340ed017d1..80bf90fa84e9a6b8d0d94198b9a0adfb3ce5e9c3 100644
--- a/plugins/rdp/rdp_plugin.c
+++ b/plugins/rdp/rdp_plugin.c
@@ -47,6 +47,7 @@
 #include "rdp_cliprdr.h"
 #include "rdp_monitor.h"
 #include "rdp_channels.h"
+#include "rdp_web_auth.h"
 
 #include <errno.h>
 #include <pthread.h>
@@ -270,7 +271,7 @@ static BOOL rf_process_event_queue(RemminaProtocolWidget *gp)
 		time(&(rfi->last_time_idle_keypress));
 		switch (event->type) {
 		case REMMINA_RDP_EVENT_TYPE_SCANCODE:
-			
+
 			if (event->key_event.extended1){
 				flags = KBD_FLAGS_EXTENDED1;
 			}
@@ -302,7 +303,7 @@ static BOOL rf_process_event_queue(RemminaProtocolWidget *gp)
 			if(rfi->clipboard.context != NULL){
 				rfi->clipboard.context->ClientFormatList(rfi->clipboard.context, event->clipboard_formatlist.pFormatList);
 			}
-			
+
 			free(event->clipboard_formatlist.pFormatList);
 			break;
 
@@ -333,7 +334,7 @@ static BOOL rf_process_event_queue(RemminaProtocolWidget *gp)
 				freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_UseMultimon, TRUE);
 				if (remmina_plugin_service->file_get_int(remminafile, "force_multimon", FALSE)) {
 					freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ForceMultimon, TRUE);
-				}	
+				}
 				freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_Fullscreen, TRUE);
 				/* got some crashes with g_malloc0, to be investigated */
 				dcml = calloc(freerdp_settings_get_uint32(rfi->clientContext.context.settings, FreeRDP_MonitorCount), sizeof(DISPLAY_CONTROL_MONITOR_LAYOUT));
@@ -1117,13 +1118,6 @@ static BOOL remmina_rdp_choose_smartcard(freerdp* instance, SmartcardCertInfo**
 	return client_cli_choose_smartcard(instance, cert_list, count, choice, gateway);
 }
 
-static BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType, char** token,
-                                    size_t count, ...)
-{
-	// TODO: Open a (currently hard coded) URL, authenticate in a webview/browser, return the access token.
-	// See client_cli_get_access_token or sdl_webview_get_access_token for implementations
-	return client_cli_get_access_token(instance, tokenType, token, count);
-}
 
 static BOOL remmina_rdp_present_gateway_message(freerdp* instance, UINT32 type, BOOL isDisplayMandatory,
                                            BOOL isConsentMandatory, size_t length,
@@ -1281,7 +1275,7 @@ static void remmina_rdp_main_loop(RemminaProtocolWidget *gp)
 		}
 		// press key(s) if we've been idle and option is selected
 		time(&cur_time);
-		time_diff_keypress = cur_time - rfi->last_time_idle_keypress;		
+		time_diff_keypress = cur_time - rfi->last_time_idle_keypress;
 		if (keypress_time > 0 && time_diff_keypress > keypress_time){
 			rfi->last_time_idle_keypress = cur_time;
 			remmina_rdp_idle_keypress(gp, &keypress_opts);
@@ -1681,28 +1675,28 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 			REMMINA_PLUGIN_DEBUG("Could not allocate assistance file structure");
 			return FALSE;
 		}
-		
-		if (remmina_plugin_service->file_get_string(remminafile, "assistance_file") == NULL || 
+
+		if (remmina_plugin_service->file_get_string(remminafile, "assistance_file") == NULL ||
 				remmina_plugin_service->file_get_string(remminafile, "assistance_pass") == NULL ){
 
 			REMMINA_PLUGIN_DEBUG("Assistance file and password are not set while assistance mode is on");
 			return FALSE;
 		}
 
-		status = freerdp_assistance_parse_file(file, 
-			remmina_plugin_service->file_get_string(remminafile, "assistance_file"), 
+		status = freerdp_assistance_parse_file(file,
+			remmina_plugin_service->file_get_string(remminafile, "assistance_file"),
 			remmina_plugin_service->file_get_string(remminafile, "assistance_pass"));
 
 		if (status < 0){
 			REMMINA_PLUGIN_DEBUG("Could not parse assistance file");
 			return FALSE;
 		}
-			
+
 
 		if (!freerdp_assistance_populate_settings_from_assistance_file(file, rfi->clientContext.context.settings)){
 			REMMINA_PLUGIN_DEBUG("Could not populate settings from assistance file");
 			return FALSE;
-		}		
+		}
 	}
 
 
@@ -1829,7 +1823,7 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 	s = remmina_plugin_service->file_get_string(remminafile, "password");
 	if (s){
 		freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_Password, s);
-	} 
+	}
 	else {
 		i = remmina_plugin_service->file_get_int(remminafile, "allow_empty_pass", 0);
 		if (i){
@@ -1838,7 +1832,7 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 		else{
 			freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_Password, s);
 		}
-		
+
 	}
 
 	freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AutoLogonEnabled, TRUE);
@@ -2083,21 +2077,39 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NlaSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ExtSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_UseRdpSecurityLayer, TRUE);
+#ifdef WITH_RDP_AUTH_AAD
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AadSecurity, FALSE);
+#endif
 	} else if (g_strcmp0(cs, "tls") == 0) {
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_RdpSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_TlsSecurity, TRUE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NlaSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ExtSecurity, FALSE);
+#ifdef WITH_RDP_AUTH_AAD
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AadSecurity, FALSE);
+#endif
 	} else if (g_strcmp0(cs, "nla") == 0) {
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_RdpSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_TlsSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NlaSecurity, TRUE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ExtSecurity, FALSE);
+#ifdef WITH_RDP_AUTH_AAD
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AadSecurity, FALSE);
+#endif
 	} else if (g_strcmp0(cs, "ext") == 0) {
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_RdpSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_TlsSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NlaSecurity, FALSE);
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ExtSecurity, TRUE);
+#ifdef WITH_RDP_AUTH_AAD
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AadSecurity, FALSE);
+	} else if (g_strcmp0(cs, "aad") == 0) {
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_RdpSecurity, FALSE);
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_TlsSecurity, FALSE);
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NlaSecurity, FALSE);
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ExtSecurity, FALSE);
+		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_AadSecurity, TRUE);
+#endif
 	} else {
 		/* This is "-nego" switch of xfreerdp */
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_NegotiateSecurityLayer, TRUE);
@@ -2350,7 +2362,7 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_UseMultimon, TRUE);
 		if (remmina_plugin_service->file_get_int(remminafile, "force_multimon", FALSE)) {
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ForceMultimon, TRUE);
-		}	
+		}
 		freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_Fullscreen, TRUE);
 
 		gchar *monitorids_string = g_strdup(remmina_plugin_service->file_get_string(remminafile, "monitorids"));
@@ -2809,7 +2821,7 @@ static void remmina_rdp_init(RemminaProtocolWidget *gp)
 		freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_AuthenticationPackageList, auth_list);
 	}
 	g_free(auth_list);
-	
+
 #endif
 	remmina_rdp_event_init(gp);
 }
@@ -2939,7 +2951,7 @@ static void remmina_rdp_call_feature(RemminaProtocolWidget *gp, const RemminaPro
 			}
 			if (remmina_plugin_service->file_get_int(remminafile, "force_multimon", FALSE)) {
 				freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_ForceMultimon, TRUE);
-			}				
+			}
 		} else {
 			REMMINA_PLUGIN_DEBUG("Remmina RDP plugin warning: Null value for rfi by REMMINA_RDP_FEATURE_MULTIMON");
 		}
@@ -3094,6 +3106,9 @@ static gpointer security_list[] =
 	"tls", N_("TLS protocol security"),
 	"rdp", N_("RDP protocol security"),
 	"ext", N_("NLA extended protocol security"),
+#ifdef WITH_RDP_AUTH_AAD
+	"aad", N_("AAD protocol security"),
+#endif
 	NULL
 };
 
@@ -3460,6 +3475,13 @@ G_MODULE_EXPORT gboolean remmina_plugin_entry(RemminaPluginService *service)
 	if (!service->register_plugin((RemminaPlugin *)&remmina_rdps))
 		return FALSE;
 
+#ifdef WITH_RDP_AUTH_AAD
+	if (!buildconfig_strstr(freerdp_get_build_config(), "WITH_AAD=ON")) {
+		REMMINA_PLUGIN_ERROR("FreeRDP is compiled without AAD authentication support");
+		return FALSE;
+	}
+#endif
+
 	if (buildconfig_strstr(freerdp_get_build_config(), "WITH_GFX_H264=ON")) {
 		gfx_h264_available = TRUE;
 		REMMINA_PLUGIN_DEBUG("gfx_h264_available: %d", gfx_h264_available);
diff --git a/plugins/rdp/rdp_web_auth.c b/plugins/rdp/rdp_web_auth.c
new file mode 100644
index 0000000000000000000000000000000000000000..14c6a1bccf14b7d3c0c88b703a1406d510970b88
--- /dev/null
+++ b/plugins/rdp/rdp_web_auth.c
@@ -0,0 +1,261 @@
+/*
+ * Remmina - The GTK+ Remote Desktop Client
+ * Copyright (C) 2010-2011 Vic Lee
+ * Copyright (C) 2014-2015 Antenore Gatta, Fabio Castelli, Giovanni Panozzo
+ * Copyright (C) 2016-2023 Antenore Gatta, Giovanni Panozzo
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ *
+ *  In addition, as a special exception, the copyright holders give
+ *  permission to link the code of portions of this program with the
+ *  OpenSSL library under certain conditions as described in each
+ *  individual source file, and distribute linked combinations
+ *  including the two.
+ *  You must obey the GNU General Public License in all respects
+ *  for all of the code used other than OpenSSL. *  If you modify
+ *  file(s) with this exception, you may extend this exception to your
+ *  version of the file(s), but you are not obligated to do so. *  If you
+ *  do not wish to do so, delete this exception statement from your
+ *  version. *  If you delete this exception statement from all source
+ *  files in the program, then also delete it here.
+ *
+ */
+
+#include "rdp_web_auth.h"
+
+#ifdef WITH_RDP_AUTH_AAD
+#include "rdp_plugin.h"
+#include <webkit2/webkit2.h>
+#include <freerdp/utils/aad.h>
+#include <winpr/wtypes.h>
+#include <winpr/wlog.h>
+
+#define AUTH_CANCELLED "__AUTH_CANCELLED__"
+
+#define SET_AUTH_URI(gp, auth_uri) \
+	g_object_set_data_full(G_OBJECT(gp), "auth-uri", auth_uri, g_free)
+
+#define GET_AUTH_URI(gp) \
+	(gchar*) g_object_get_data(G_OBJECT(gp), "auth-uri")
+
+#define SET_TOKEN_URI(gp, token_uri) \
+	g_object_set_data_full(G_OBJECT(gp), "token-uri", token_uri, g_free)
+
+#define GET_TOKEN_URI(gp) \
+	(gchar*) g_object_get_data(G_OBJECT(gp), "token-uri")
+
+static char* extract_authorization_code(char* url)
+{
+	WINPR_ASSERT(url);
+
+	for (char* p = strchr(url, '?'); p++ != NULL; p = strchr(p, '&'))
+	{
+		if (strncmp(p, "code=", 5) != 0)
+			continue;
+
+		char* end = NULL;
+		p += 5;
+
+		end = strchr(p, '&');
+		if (end)
+			*end = '\0';
+
+		return p;
+	}
+
+	return NULL;
+}
+
+static void delete_event_cb(GtkWidget *dialog, GdkEvent* event, RemminaProtocolWidget *gp) {
+	SET_TOKEN_URI(gp, AUTH_CANCELLED);
+}
+
+static gboolean decide_policy_cb(WebKitWebView *web_view, WebKitPolicyDecision *decision,
+	WebKitPolicyDecisionType type, RemminaProtocolWidget *gp)
+{
+	WebKitNavigationPolicyDecision *nav_decision;
+
+	switch (type) {
+	case WEBKIT_POLICY_DECISION_TYPE_NAVIGATION_ACTION:
+			nav_decision = WEBKIT_NAVIGATION_POLICY_DECISION(decision);
+			WebKitNavigationAction *nav_action =
+				webkit_navigation_policy_decision_get_navigation_action(nav_decision);
+
+		if (webkit_navigation_action_is_redirect(nav_action)) {
+			WebKitURIRequest *req = webkit_navigation_action_get_request(nav_action);
+			const gchar *uri = webkit_uri_request_get_uri(req);
+			SET_TOKEN_URI(gp, (void *)uri);
+			GtkWidget *dialog = gtk_widget_get_toplevel(GTK_WIDGET(web_view));
+			gtk_widget_destroy(GTK_WIDGET(dialog));
+		}
+		return FALSE;
+	default:
+		/* Making no decision results in webkit_policy_decision_use(). */
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static BOOL remmina_rdp_webview_show(RemminaProtocolWidget *gp) {
+	char* auth_uri = GET_AUTH_URI(gp);
+
+	GtkWidget *dialog = gtk_window_new(GTK_WINDOW_TOPLEVEL);
+	gtk_window_set_default_size(GTK_WINDOW(dialog), 500, 500);
+
+	WebKitWebView *webView = WEBKIT_WEB_VIEW(webkit_web_view_new());
+	gtk_container_add(GTK_CONTAINER(dialog), GTK_WIDGET(webView));
+
+	webkit_web_view_load_uri(webView, auth_uri);
+
+	g_signal_connect(dialog, "delete-event", G_CALLBACK(delete_event_cb), gp);
+	g_signal_connect(webView, "decide-policy", G_CALLBACK(decide_policy_cb), gp);
+
+	gtk_widget_show_all(dialog);
+
+	return G_SOURCE_REMOVE;
+}
+
+static BOOL remmina_rdp_get_rdsaad_access_token(freerdp* instance, const char* scope,
+	const char* req_cnf, char** token)
+{
+	assert(instance);
+	assert(instance->context);
+
+	rfContext *rfi;
+	RemminaProtocolWidget *gp;
+
+	rfi = (rfContext *)instance->context;
+	gp = rfi->protocol_widget;
+
+	size_t size = 0;
+	char* response_uri = NULL;
+	char* token_request = NULL;
+	char* redirect_uri = NULL;
+	size_t redirect_uri_len = 0;
+	char* auth_uri = NULL;
+	size_t auth_uri_len = 0;
+
+	assert(scope);
+	assert(req_cnf);
+	assert(token);
+
+	BOOL rc = FALSE;
+	*token = NULL;
+
+	const char* client_id =
+	    freerdp_settings_get_string(instance->context->settings, FreeRDP_GatewayAvdClientID);
+	if (!client_id)
+		goto cleanup;
+
+	winpr_asprintf(&redirect_uri, &redirect_uri_len,
+	               "ms-appx-web%%3a%%2f%%2fMicrosoft.AAD.BrokerPlugin%%2f%s", client_id);
+	if (!redirect_uri)
+		goto cleanup;
+
+	const char* ep = freerdp_utils_aad_get_wellknown_string(instance->context,
+	                                                        AAD_WELLKNOWN_authorization_endpoint);
+
+	winpr_asprintf(&auth_uri, &auth_uri_len, "%s?client_id=%s&response_type="
+	       "code&scope=%s&redirect_uri=%s",
+	       ep, client_id, scope, redirect_uri);
+	if (!auth_uri)
+	{
+		goto cleanup;
+	}
+	SET_AUTH_URI(gp, auth_uri);
+	IDLE_ADD((GSourceFunc)remmina_rdp_webview_show, gp);
+
+	char* token_uri;
+	while(true) {
+		usleep(500 * 1000);
+		token_uri = GET_TOKEN_URI(gp);
+		if (token_uri == NULL)
+		{
+			continue;
+		}
+
+		if (g_str_equal(token_uri, AUTH_CANCELLED))
+		{
+			rc = FALSE;
+			goto cleanup;
+		}
+
+		char* code = extract_authorization_code(token_uri);
+		if (!code) {
+			goto cleanup;
+		}
+
+		if (winpr_asprintf(&token_request, &size,
+						"grant_type=authorization_code&code=%s&client_id=%s&scope=%s&redirect_uri=%"
+						"s&req_cnf=%s",
+						code, client_id, scope, redirect_uri, req_cnf) <= 0)
+		{
+			goto cleanup;
+		}
+
+		rc = client_common_get_access_token(instance, token_request, token);
+		break;
+	}
+
+cleanup:
+	free(auth_uri);
+	free(redirect_uri);
+	free(token_request);
+	free(response_uri);
+	return rc && (*token != NULL);
+}
+
+BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType, char** token,
+									size_t count, ...)
+{
+	switch (tokenType)
+	{
+		case ACCESS_TOKEN_TYPE_AAD:
+		{
+			if (count < 2)
+			{
+				REMMINA_PLUGIN_ERROR(
+				         "ACCESS_TOKEN_TYPE_AAD expected 2 additional arguments, but got %zu, aborting",
+				         count);
+				return FALSE;
+			}
+			else if (count > 2)
+				REMMINA_PLUGIN_WARNING(
+				          "ACCESS_TOKEN_TYPE_AAD expected 2 additional arguments, but got %zu, ignoring",
+				          count);
+			va_list ap = { 0 };
+			va_start(ap, count);
+			const char* scope = va_arg(ap, const char*);
+			const char* req_cnf = va_arg(ap, const char*);
+			const BOOL rc = remmina_rdp_get_rdsaad_access_token(instance, scope, req_cnf, token);
+			va_end(ap);
+			return rc;
+		}
+		// TODO: case ACCESS_TOKEN_TYPE_AVD:
+		default:
+			REMMINA_PLUGIN_ERROR("Unexpected value for AccessTokenType [%" PRIuz "], aborting", tokenType);
+			return FALSE;
+	}
+}
+#else
+#if FREERDP_VERSION_MAJOR >= 3
+BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType, char** token,
+                                    size_t count, ...)
+{
+	return client_cli_get_access_token(instance, tokenType, token, count);
+}
+#endif
+#endif
diff --git a/plugins/rdp/rdp_web_auth.h b/plugins/rdp/rdp_web_auth.h
new file mode 100644
index 0000000000000000000000000000000000000000..2b0f1dd76003ef6f01f7a55c1d293554748f81b8
--- /dev/null
+++ b/plugins/rdp/rdp_web_auth.h
@@ -0,0 +1,44 @@
+/*
+ * Remmina - The GTK+ Remote Desktop Client
+ * Copyright (C) 2010-2011 Vic Lee
+ * Copyright (C) 2017-2023 Antenore Gatta, Giovanni Panozzo
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA  02110-1301, USA.
+ *
+ *  In addition, as a special exception, the copyright holders give
+ *  permission to link the code of portions of this program with the
+ *  OpenSSL library under certain conditions as described in each
+ *  individual source file, and distribute linked combinations
+ *  including the two.
+ *  You must obey the GNU General Public License in all respects
+ *  for all of the code used other than OpenSSL. *  If you modify
+ *  file(s) with this exception, you may extend this exception to your
+ *  version of the file(s), but you are not obligated to do so. *  If you
+ *  do not wish to do so, delete this exception statement from your
+ *  version. *  If you delete this exception statement from all source
+ *  files in the program, then also delete it here.
+ *
+ */
+
+#pragma once
+
+#include <freerdp/version.h>
+#if FREERDP_VERSION_MAJOR >= 3
+#include <freerdp/freerdp.h>
+
+BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType, char** token,
+                                    size_t count, ...);
+#endif
