diff --git a/plugins/rdp/rdp_file.c b/plugins/rdp/rdp_file.c
index e43cfb92937718f249ae89b9d4b61fbe8e2523e5..e93a6aef7d15b62c32ee3304e6b0d9445adbd21b 100644
--- a/plugins/rdp/rdp_file.c
+++ b/plugins/rdp/rdp_file.c
@@ -56,6 +56,12 @@ gboolean remmina_rdp_file_import_test(RemminaFilePlugin *plugin, const gchar *fr
 	if (g_strcmp0(ext, "rdp") == 0)
 		return TRUE;
 
+	if (g_strcmp0(ext, "RDPW") == 0)
+		return TRUE;
+
+	if (g_strcmp0(ext, "rdpw") == 0)
+		return TRUE;
+
 	return FALSE;
 }
 
@@ -101,6 +107,8 @@ static void remmina_rdp_file_import_field(RemminaFile *remminafile, const gchar
 		remmina_plugin_service->file_set_string(remminafile, "gatewayaccesstoken", value);
 	} else if (g_strcmp0(key, "authentication level") == 0) {
 		remmina_plugin_service->file_set_int(remminafile, "authentication level", atoi(value));
+	} else if (g_strcmp0(key, "remoteapplicationprogram") == 0) {
+		remmina_plugin_service->file_set_string(remminafile, "remoteapplicationprogram", value);
 	}
 	/* tsclient fields, import only */
 	else if (g_strcmp0(key, "client hostname") == 0) {
diff --git a/plugins/rdp/rdp_plugin.c b/plugins/rdp/rdp_plugin.c
index 80bf90fa84e9a6b8d0d94198b9a0adfb3ce5e9c3..2e9cc76f824df34129176d45444a9e1a7e2cd941 100644
--- a/plugins/rdp/rdp_plugin.c
+++ b/plugins/rdp/rdp_plugin.c
@@ -1894,12 +1894,27 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 		if (g_strcmp0(cs, "http") == 0) {
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayRpcTransport, FALSE);
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpTransport, TRUE);
+#ifdef WITH_RDP_AUTH_AAD
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayArmTransport, FALSE);
+#endif
 		} else if (g_strcmp0(cs, "rpc") == 0) {
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayRpcTransport, TRUE);
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpTransport, FALSE);
+#ifdef WITH_RDP_AUTH_AAD
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpUseWebsockets, FALSE);
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayArmTransport, FALSE);
+#endif
 		} else if (g_strcmp0(cs, "auto") == 0) {
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayRpcTransport, TRUE);
 			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpTransport, TRUE);
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayArmTransport, FALSE);
+#ifdef WITH_RDP_AUTH_AAD
+		} else if (g_strcmp0(cs, "arm") == 0) {
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayRpcTransport, FALSE);
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpTransport, FALSE);
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayHttpUseWebsockets, FALSE);
+			freerdp_settings_set_bool(rfi->clientContext.context.settings, FreeRDP_GatewayArmTransport, TRUE);
+#endif
 		}
 		remmina_plugin_service->get_server_port(s, 443, &gateway_host, &gateway_port);
 		freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_GatewayHostname, gateway_host);
@@ -1966,6 +1981,10 @@ static gboolean remmina_rdp_main(RemminaProtocolWidget *gp)
 		freerdp_settings_set_pointer_len(rfi->clientContext.context.settings, FreeRDP_LoadBalanceInfo, tmp, strlen(tmp));
 	}
 
+	if (remmina_plugin_service->file_get_string(remminafile, "remoteapplicationprogram")) {
+		freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_RemoteApplicationProgram, remmina_plugin_service->file_get_string(remminafile, "remoteapplicationprogram"));
+	}
+
 	if (remmina_plugin_service->file_get_string(remminafile, "exec"))
 		freerdp_settings_set_string(rfi->clientContext.context.settings, FreeRDP_AlternateShell, remmina_plugin_service->file_get_string(remminafile, "exec"));
 
@@ -3145,6 +3164,9 @@ static gpointer gwtransp_list[] =
 {
 	"http", "HTTP",
 	"rpc",	"RPC",
+#ifdef WITH_RDP_AUTH_AAD
+	"arm", "ARM",
+#endif
 	"auto", "Auto",
 	NULL
 };
@@ -3300,6 +3322,7 @@ static const RemminaProtocolSetting remmina_rdp_advanced_settings[] =
 	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "exec",		    N_("Start-up program"),				 FALSE, NULL,		  NULL														 },
 	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "execpath",		    N_("Start-up path"),				 FALSE, NULL,		  NULL														 },
 	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "loadbalanceinfo",	    N_("Load balance info"),				 FALSE, NULL,		  NULL														 },
+	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "remoteapplicationprogram", N_("Remote Application Program"), 		 FALSE, NULL,		  NULL														 },
 	// TRANSLATORS: Do not use typographic quotation marks, these must stay as "double quote", also know as “Typewriter ("programmer's") quote, ambidextrous.”
 	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "printer_overrides",	    N_("Override printer drivers"),			 FALSE, NULL,		  N_("\"Samsung_CLX-3300_Series\":\"Samsung CLX-3300 Series PS\";\"Canon MF410\":\"Canon MF410 Series UFR II\"") },
 	{ REMMINA_PROTOCOL_SETTING_TYPE_TEXT,	  "usb",		    N_("USB device redirection"),			 TRUE,	NULL,		  usb_tooltip													 },
diff --git a/plugins/rdp/rdp_web_auth.c b/plugins/rdp/rdp_web_auth.c
index 14c6a1bccf14b7d3c0c88b703a1406d510970b88..41b35f384b43352a79733d3e8a52bd77e178794e 100644
--- a/plugins/rdp/rdp_web_auth.c
+++ b/plugins/rdp/rdp_web_auth.c
@@ -46,13 +46,13 @@
 #define AUTH_CANCELLED "__AUTH_CANCELLED__"
 
 #define SET_AUTH_URI(gp, auth_uri) \
-	g_object_set_data_full(G_OBJECT(gp), "auth-uri", auth_uri, g_free)
+	g_object_set_data(G_OBJECT(gp), "auth-uri", auth_uri)
 
 #define GET_AUTH_URI(gp) \
 	(gchar*) g_object_get_data(G_OBJECT(gp), "auth-uri")
 
 #define SET_TOKEN_URI(gp, token_uri) \
-	g_object_set_data_full(G_OBJECT(gp), "token-uri", token_uri, g_free)
+	g_object_set_data(G_OBJECT(gp), "token-uri", token_uri)
 
 #define GET_TOKEN_URI(gp) \
 	(gchar*) g_object_get_data(G_OBJECT(gp), "token-uri")
@@ -140,9 +140,8 @@ static BOOL remmina_rdp_get_rdsaad_access_token(freerdp* instance, const char* s
 	rfi = (rfContext *)instance->context;
 	gp = rfi->protocol_widget;
 
-	size_t size = 0;
-	char* response_uri = NULL;
 	char* token_request = NULL;
+	size_t token_request_len = 0;
 	char* redirect_uri = NULL;
 	size_t redirect_uri_len = 0;
 	char* auth_uri = NULL;
@@ -171,10 +170,106 @@ static BOOL remmina_rdp_get_rdsaad_access_token(freerdp* instance, const char* s
 	winpr_asprintf(&auth_uri, &auth_uri_len, "%s?client_id=%s&response_type="
 	       "code&scope=%s&redirect_uri=%s",
 	       ep, client_id, scope, redirect_uri);
-	if (!auth_uri)
-	{
+	if (!auth_uri) {
 		goto cleanup;
 	}
+
+	SET_AUTH_URI(gp, auth_uri);
+	IDLE_ADD((GSourceFunc)remmina_rdp_webview_show, gp);
+
+	char* token_uri;
+	while(true) {
+		usleep(500 * 1000);
+		token_uri = GET_TOKEN_URI(gp);
+		if (token_uri == NULL) {
+			continue;
+		}
+
+		if (g_str_equal(token_uri, AUTH_CANCELLED)) {
+			rc = FALSE;
+			goto cleanup;
+		}
+
+		char* code = extract_authorization_code(token_uri);
+		if (!code) {
+			goto cleanup;
+		}
+
+		if (winpr_asprintf(&token_request, &token_request_len,
+						"grant_type=authorization_code&code=%s&client_id=%s&scope=%s&redirect_uri=%"
+						"s&req_cnf=%s",
+						code, client_id, scope, redirect_uri, req_cnf) <= 0)
+		{
+			goto cleanup;
+		}
+
+		rc = client_common_get_access_token(instance, token_request, token);
+		break;
+	}
+
+cleanup:
+	free(auth_uri);
+	free(redirect_uri);
+	free(token_request);
+	SET_AUTH_URI(gp, NULL);
+	SET_TOKEN_URI(gp, NULL);
+	return rc && (*token != NULL);
+}
+
+static BOOL remmina_rdp_get_avd_access_token(freerdp* instance, char** token) {
+	assert(instance);
+	assert(instance->context);
+
+	rfContext *rfi;
+	RemminaProtocolWidget *gp;
+
+	rfi = (rfContext *)instance->context;
+	gp = rfi->protocol_widget;
+
+	char* token_request = NULL;
+	size_t token_request_len = 0;
+	char* redirect_uri = NULL;
+	size_t redirect_uri_len = 0;
+	char* auth_uri = NULL;
+	size_t auth_uri_len = 0;
+	const char* scope = "https%3A%2F%2Fwww.wvd.microsoft.com%2F.default";
+
+	assert(token);
+
+	BOOL rc = FALSE;
+	*token = NULL;
+
+	const char* client_id =
+	    freerdp_settings_get_string(instance->context->settings, FreeRDP_GatewayAvdClientID);
+	const char* base = freerdp_settings_get_string(instance->context->settings,
+	                                               FreeRDP_GatewayAzureActiveDirectory);
+	const BOOL useTenant =
+	    freerdp_settings_get_bool(instance->context->settings, FreeRDP_GatewayAvdUseTenantid);
+	const char* tenantid = "common";
+	if (useTenant) {
+		tenantid =
+		    freerdp_settings_get_string(instance->context->settings, FreeRDP_GatewayAvdAadtenantid);
+	}
+	if (!base || !tenantid || !client_id) {
+		goto cleanup;
+	}
+
+	winpr_asprintf(&redirect_uri, &redirect_uri_len,
+	               "https%%3A%%2F%%2F%s%%2F%s%%2Foauth2%%2Fnativeclient", base, tenantid);
+	if (!redirect_uri) {
+		goto cleanup;
+	}
+
+	const char* ep = freerdp_utils_aad_get_wellknown_string(instance->context,
+	                                                        AAD_WELLKNOWN_authorization_endpoint);
+
+	winpr_asprintf(&auth_uri, &auth_uri_len, "%s?client_id=%s&response_type="
+	       "code&scope=%s&redirect_uri=%s",
+	       ep, client_id, scope, redirect_uri);
+	if (!auth_uri) {
+		goto cleanup;
+	}
+
 	SET_AUTH_URI(gp, auth_uri);
 	IDLE_ADD((GSourceFunc)remmina_rdp_webview_show, gp);
 
@@ -198,10 +293,10 @@ static BOOL remmina_rdp_get_rdsaad_access_token(freerdp* instance, const char* s
 			goto cleanup;
 		}
 
-		if (winpr_asprintf(&token_request, &size,
-						"grant_type=authorization_code&code=%s&client_id=%s&scope=%s&redirect_uri=%"
-						"s&req_cnf=%s",
-						code, client_id, scope, redirect_uri, req_cnf) <= 0)
+		if (winpr_asprintf(
+				&token_request, &token_request_len,
+				"grant_type=authorization_code&code=%s&client_id=%s&scope=%s&redirect_uri=%s", code,
+				client_id, scope, redirect_uri) <= 0)
 		{
 			goto cleanup;
 		}
@@ -214,7 +309,8 @@ cleanup:
 	free(auth_uri);
 	free(redirect_uri);
 	free(token_request);
-	free(response_uri);
+	SET_AUTH_URI(gp, NULL);
+	SET_TOKEN_URI(gp, NULL);
 	return rc && (*token != NULL);
 }
 
@@ -232,10 +328,11 @@ BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType,
 				         count);
 				return FALSE;
 			}
-			else if (count > 2)
+			else if (count > 2) {
 				REMMINA_PLUGIN_WARNING(
 				          "ACCESS_TOKEN_TYPE_AAD expected 2 additional arguments, but got %zu, ignoring",
 				          count);
+			}
 			va_list ap = { 0 };
 			va_start(ap, count);
 			const char* scope = va_arg(ap, const char*);
@@ -244,7 +341,15 @@ BOOL remmina_rdp_get_access_token(freerdp* instance, AccessTokenType tokenType,
 			va_end(ap);
 			return rc;
 		}
-		// TODO: case ACCESS_TOKEN_TYPE_AVD:
+		case ACCESS_TOKEN_TYPE_AVD:
+		{
+			if (count != 0) {
+				REMMINA_PLUGIN_WARNING(
+				          "ACCESS_TOKEN_TYPE_AVD expected 0 additional arguments, but got %zu, ignoring",
+				          count);
+			}
+			return remmina_rdp_get_avd_access_token(instance, token);
+		}
 		default:
 			REMMINA_PLUGIN_ERROR("Unexpected value for AccessTokenType [%" PRIuz "], aborting", tokenType);
 			return FALSE;
diff --git a/src/remmina_main.c b/src/remmina_main.c
index 554f616c9ded2deeb0e1ef840c85b9ca1376908c..5e75724fecae08cbe2bcd962421fbbc0d24104af 100644
--- a/src/remmina_main.c
+++ b/src/remmina_main.c
@@ -1379,8 +1379,10 @@ static void remmina_set_file_chooser_filters(GtkFileChooser *chooser)
 	filter = gtk_file_filter_new();
 	gtk_file_filter_set_name(filter, _("RDP Files"));
 	gtk_file_filter_add_pattern(filter, "*.rdp");
+	gtk_file_filter_add_pattern(filter, "*.rdpw");
 	gtk_file_filter_add_pattern(filter, "*.rdpx");
 	gtk_file_filter_add_pattern(filter, "*.RDP");
+	gtk_file_filter_add_pattern(filter, "*.RDPW");
 	gtk_file_filter_add_pattern(filter, "*.RDPX");
 	gtk_file_chooser_add_filter(GTK_FILE_CHOOSER(chooser), filter);
 	gtk_file_chooser_set_filter(GTK_FILE_CHOOSER(chooser), filter);
