--- a/compiler/rustc_target/src/spec/aarch64_unknown_linux_musl.rs	2020-12-29 04:03:08.000000000 +0100
+++ b/compiler/rustc_target/src/spec/aarch64_unknown_linux_musl.rs	2020-12-31 19:13:13.588052924 +0100
@@ -3,6 +3,7 @@
 pub fn target() -> Target {
     let mut base = super::linux_musl_base::opts();
     base.max_atomic_width = Some(128);
+    base.static_position_independent_executables = true;
 
     Target {
         llvm_target: "aarch64-unknown-linux-musl".to_string(),
