--- a/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs	2025-02-21 10:09:22.256865952 +0100
+++ b/compiler/rustc_target/src/spec/targets/aarch64_unknown_linux_musl.rs	2025-02-21 10:08:54.341173464 +0100
@@ -6,6 +6,7 @@
     base.supports_xray = true;
     base.features = "+v8a".into();
     base.stack_probes = StackProbeType::Inline;
+    base.static_position_independent_executables = true;
     base.supported_sanitizers = SanitizerSet::ADDRESS
         | SanitizerSet::CFI
         | SanitizerSet::LEAK
