--- a/Directory.Build.props	2025-09-13 19:55:38.448826903 +0200
+++ b/Directory.Build.props	2025-09-13 19:55:47.332737429 +0200
@@ -12,7 +12,7 @@
     <RepositoryBranch>master</RepositoryBranch>
 
     <!-- Set target .NET version. -->
-    <TargetFramework>net6.0</TargetFramework>
+    <TargetFramework>net8.0</TargetFramework>
     <RollForward>Major</RollForward>
 
     <!-- Deterministic builds. -->
